src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,221 @@
+            return;
+        } else {
+            newNode = new Node(value);
+            wire(target.prev, newNode);
+            wire(newNode, target);
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T set(T value,int index) {
+        validateIndex(index);
+        Node target = getNodeByIndex(index);
+        T result = target.value;
+        target.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node target = getNodeByIndex(index);
+        T result = target.value;
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (size == 2) {
+            if (target.value.equals(head.value)) {
+                head = tail;
+            } else {
+                tail = head;
+            }
+            wire(head, tail);
+        } else if (index == 0) {
+            head = target.next;
+            head.prev = head;
+        } else if (index == size - 1) {
+            tail = target.prev;
+            tail.next = null;
+        } else {
+            wire(target.prev, target.next);
+        }
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T t) {
-        return false;
+        Node target = getNodeByValue(t);
+        if (target == null) {
+            return false;
+        }
+
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (size == 2) {
+            if (target.value == head.value || target.value.equals(head.value)) {
+                head = tail;
+            } else {
+                tail = head;
+            }
+            wire(head, tail);
+            release(target);
+        } else if (head.value == t || head.value.equals(t)) {
+            head = target.next;
+            head.prev = head;
+            release(target);
+        } else {
+            if (target != tail) {
+                wire(target.prev, target.next);
+            } else if (tail.value == t || tail.value.equals(t)) {
+                tail = target.prev;
+                tail.next = null;
+            }
+            release(target);
+        }
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException(index);
```

Comment:
let's add message to make exception more informative

Reply:
integer value will be wrapped in default message, which is used across whole jdk libraries